controllers/secretdefinition_controller.go

```diff
@@ -0,0 +1,242 @@
+		decoder, err := backend.NewDecoder(v.Encoding)
+		if err != nil {
+			r.Log.Error(err, "refusing to use encoding", "encoding", v.Encoding)
+			return nil, err
+		}
+		desiredState[k], err = decoder.DecodeString(bSecret)
+		if err != nil {
+			r.Log.Error(err, "unable to decode data for secret", "encoding", v.Encoding, "path", v.Path, "key", v.Key)
+			return nil, err
+		}
+	}
+	return desiredState, err
+}
+
+// getCurrentState reads the content from the Kubernetes Secret API object for later comparison
+func (r *SecretDefinitionReconciler) getCurrentState(namespace string, name string) (map[string][]byte, error) {
+	// We don't read secrets from cache, as it's not the object we reconcile
+	reader := r.APIReader
+	data := make(map[string][]byte)
+	secret := &corev1.Secret{}
+	err := reader.Get(r.Ctx, client.ObjectKey{
+		Namespace: namespace,
+		Name:      name,
+	}, secret)
+	if err != nil {
+		secretReadErrorsTotal.WithLabelValues(name, namespace).Inc()
+		return data, err
+	}
+	data = secret.Data
+	return data, err
+}
+
+// upsertSecret will create or update a secret
+func (r *SecretDefinitionReconciler) upsertSecret(secretDef *secretsmanagerv1alpha1.SecretDefinition, data map[string][]byte) error {
+	secret := &corev1.Secret{
+		Type: corev1.SecretType(secretDef.Spec.Type),
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: secretDef.Namespace,
+			Labels: map[string]string{
+				"managedBy":     "secrets-manager",
+				"lastUpdatedAt": time.Now().Format(timestampFormat),
+			},
+			Name: secretDef.Spec.Name,
+		},
+		Data: data,
+	}
+	err := r.Create(r.Ctx, secret)
+	if errors.IsAlreadyExists(err) {
+		err = r.Update(r.Ctx, secret)
+	}
+	return err
+}
+
+// deleteSecret will delete a secret given its namespace and name
+func (r *SecretDefinitionReconciler) deleteSecret(namespace string, name string) error {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      name,
+		},
+	}
+	return r.Delete(r.Ctx, secret)
+}
+
+// +kubebuilder:rbac:groups=secrets-manager.tuenti.io,resources=secretdefinitions,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=secrets-manager.tuenti.io,resources=secretdefinitions/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;patch;delete
+func (r *SecretDefinitionReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("secretdefinition", req.NamespacedName)
+
+	secretDef := &secretsmanagerv1alpha1.SecretDefinition{}
+
+	err := r.Get(r.Ctx, req.NamespacedName, secretDef)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("could not get SecretDefinition '%s'", req.NamespacedName))
+		return ctrl.Result{}, ignoreNotFoundError(err)
+	}
+
+	secretName := secretDef.Spec.Name
+	secretNamespace := secretDef.Namespace
+
+	log = log.WithValues("secret", fmt.Sprintf("%s/%s", secretNamespace, secretName))
+
+	if secretDef.ObjectMeta.DeletionTimestamp.IsZero() {
```

Comment:
(suggestion - minor) name the condition (extract to an aux function)

```go
if secretDef.IsMarkedForRemoval() { ...

---

func (r *SecretDefinition) IsMarkedForRemoval() bool {
    return r.ObjectMeta.DeletionTimestamp.IsZero()
}
```

Reply:
makes sense, thanks @manutero